### ISSUE TITLE
How to use bolt to optimize bin continuously in the production environment

### ISSUE THREAD

**Issue by zpget** [facebookincubator/BOLT#94]
use bolt to optimize bin continuously in the production environment, find a problem
step 1. bin1 -> bolt ->bin1.blot
step 2. use bin1.bolt in the production environment,  collect this bin1 profile data （Make full use of the production environment）
step 3 use bin1 profile data to optimize new compile bin2 (code difference is relatively small)
There is no optimization effect, so  how to applay to realy production environment ?
BOLT-INFO: program-wide dynostats after all optimizations before SCTC and FOP (no change):

                   0 : executed forward branches
                   0 : taken forward branches
                   0 : executed backward branches
                   0 : taken backward branches
                   0 : executed unconditional branches
                   0 : all function calls
                   0 : indirect calls
                   0 : PLT calls
                   0 : executed instructions
                   0 : executed load instructions
                   0 : executed store instructions
                   0 : taken jump table branches
                   0 : taken unknown indirect branches
                   0 : total branches
                   0 : taken branches
                   0 : non-taken conditional branches
                   0 : taken conditional branches
                   0 : all conditional branches

**Comment by maksfb** [comment]
If you want to collect a profile on a BOLTed binary, you need to add `-enable-bat` option to BOLT invocation. This will add address-translation tables into the binary that will allow `perf2bolt` to correctly attribute events to addresses in the original binary. Note that the option will increase the binary size, but only on disk, not in memory.

**Comment by zpget** [comment]
use -enable-bat to gen bin.bolt
use 
1. bin.bolt to collect profile data
2. then trans profile data with original bin , llvm-bolt to new bin2.bolt
but still get this info:
BOLT-INFO: program-wide dynostats after all optimizations before SCTC and FOP (no change):

                   0 : executed forward branches
                   0 : taken forward branches
                   0 : executed backward branches
                   0 : taken backward branches
                   0 : executed unconditional branches
                   0 : all function calls
                   0 : indirect calls
                   0 : PLT calls
                   0 : executed instructions
                   0 : executed load instructions
                   0 : executed store instructions
                   0 : taken jump table branches
                   0 : taken unknown indirect branches
                   0 : total branches
                   0 : taken branches
                   0 : non-taken conditional branches
                   0 : taken conditional branches
                   0 : all conditional branches

**Comment by maksfb** [comment]
You have to pass `bin.bolt` to `perf2bolt` which will use the address translation tables for profile conversion.  Then use the original binary when running `llvm-bolt`.

**Comment by zpget** [comment]
use bin.bolt to perf2bolt
perf2bolt -p bin.bolt_perf.data -o bin.bolt_perf.fdata bin.bolt -w bin.bolt.yaml

llvm-bolt -enable-bat=1 bin -o bin.bolt -b=bin.yaml -align-macro-fusion=all -reorder-blocks=cache+ -reorder-functions=hfsort+ -split-functions=3 -split-all-cold -split-eh -dyno-stats -icf=1 -update-debug-sections

get many warings and still as this:

BOLT-INFO: profile for 115 objects was ignored
BOLT-INFO: the input contains 5809 (dynamic count : 0) opportunities for macro-fusion optimization that are going to be fixed
BOLT-INFO: removed 264 'repz' prefixes with estimated execution count of 0 times.
BOLT-INFO: ICF folded 19356 out of 69534 functions in 13 passes. 1 functions had jump tables.
BOLT-INFO: Removing all identical functions will save 1391.48 KB of code space. Folded functions were called 0 times based on profile.
BOLT-INFO: basic block reordering modified layout of 0 (0.00%) functions
BOLT-INFO: UCE removed 0 blocks and 0 bytes of code.
BOLT-INFO: running hfsort+ for 577 functions
BOLT-INFO: program-wide dynostats after all optimizations before SCTC and FOP (no change):

                   0 : executed forward branches
                   0 : taken forward branches
                   0 : executed backward branches
                   0 : taken backward branches
                   0 : executed unconditional branches
                   0 : all function calls
                   0 : indirect calls
                   0 : PLT calls
                   0 : executed instructions
                   0 : executed load instructions
                   0 : executed store instructions
                   0 : taken jump table branches
                   0 : taken unknown indirect branches
                   0 : total branches
                   0 : taken branches
                   0 : non-taken conditional branches
                   0 : taken conditional branches
                   0 : all conditional branches

**Comment by maksfb** [comment]
What's the output from `perf2bolt`?

**Comment by zpget** [comment]
PERF2BOLT: spawning perf job to read branch events
PERF2BOLT: spawning perf job to read mem events
PERF2BOLT: spawning perf job to read process events
PERF2BOLT: spawning perf job to read task events
BOLT-INFO: Target architecture: x86_64
BOLT-INFO: first alloc address is 0x400000
BOLT-INFO: creating new program header table at address 0x6000000, offset 0x5c00000
BOLT-INFO: Parsed 50616 BAT entries
BOLT-INFO: Parsed 438 BAT cold-to-hot entries
BOLT-INFO: disabling -align-macro-fusion in non-relocation mode
BOLT-INFO: pre-processing profile using perf data aggregator
BOLT-INFO: profile collection done on a binary already processed by BOLT
BOLT-WARNING: build-id will not be checked because we could not read one from input binary
PERF2BOLT: waiting for perf mmap events collection to finish...
wait result:0 2
wait end result:0
PERF2BOLT: parsing perf-script mmap events output
PERF2BOLT: waiting for perf task events collection to finish...
wait result:0 2
wait end result:0
PERF2BOLT: parsing perf-script task events output
PERF2BOLT: input binary is associated with 1 PID(s)
PERF2BOLT: waiting for perf events collection to finish...
wait result:0 2
wait end result:0
PERF2BOLT: parse branch events...
PERF2BOLT: read 9621 samples and 148926 LBR entries
PERF2BOLT: 0 samples (0.0%) were ignored
PERF2BOLT: traces mismatching disassembled function contents: 171 (0.1%)
PERF2BOLT: out of range traces involving unknown regions: 69345 (48.6%)

**Comment by maksfb** [comment]
Does it say how many objects were written to `bin.bolt_perf.fdata`? Also, does `llvm-bolt` log say anything about stale functions?

**Comment by zpget** [comment]
PERF2BOLT: processing branch events...
PERF2BOLT: wrote 6347 objects and 0 memory objects to bin.bolt_perf.fdata

BOLT-WARNING: 686 (100.0% of all profiled) functions have invalid (possibly stale) profile. Use -report-stale to see the list.

**Comment by zpget** [comment]
so collect a profile on a BOLTed binary may have same problems,  and must need offline original bin environment to collect profile ? how facebook date center ongoing optimization ？

**Comment by zpget** [comment]
> Does it say how many objects were written to `bin.bolt_perf.fdata`? Also, does `llvm-bolt` log say anything about stale functions?

bin code same with bin.bolt

perf2bot:
PERF2BOLT: processing branch events...
PERF2BOLT: wrote 6347 objects and 0 memory objects to bin.bolt_perf.fdata

llvm-bolt:
BOLT-WARNING: 686 (100.0% of all profiled) functions have invalid (possibly stale) profile. Use -report-stale to see the list.

**Comment by zpget** [comment]
so collect a profile on a BOLTed binary may have same problems, and must need offline original bin environment to collect profile ? how facebook date center ongoing optimization ？

**Comment by maksfb** [comment]
I suspect I know what's going on. I'll put up a fix shortly for you to try.

**Comment by zpget** [comment]
> I suspect I know what's going on. I'll put up a fix shortly for you to try.

RE:Fix for SKL bug workaround 
update code ,test  result still as this, so many warnings and SCTC and FOP (no change) results
=====================================================
BOLT-WARNING: profile ignored for function _ZSt16__introsort_loopIN9__gnu_cxx17__normal_iteratorIPSsSt6vectorISsSaISsEEEElNS0_5__ops15_Iter_less_iterEEvT_S9_T0_T1_.isra.64.cold.0/1(*4)
xxxx
xxx
BOLT-INFO: 597 out of 68829 functions in the binary (0.9%) have non-empty execution profile
BOLT-INFO: 5 functions with profile could not be optimized
BOLT-WARNING: 597 (100.0% of all profiled) functions have invalid (possibly stale) profile. Use -report-stale to see the list.
BOLT-INFO: profile for 94 objects was ignored
BOLT-INFO: the input contains 5809 (dynamic count : 0) opportunities for macro-fusion optimization that are going to be fixed
BOLT-INFO: removed 264 'repz' prefixes with estimated execution count of 0 times.
BOLT-INFO: ICF folded 19356 out of 69534 functions in 12 passes. 1 functions had jump tables.
BOLT-INFO: Removing all identical functions will save 1391.48 KB of code space. Folded functions were called 0 times based on profile.
BOLT-INFO: basic block reordering modified layout of 0 (0.00%) functions
BOLT-INFO: UCE removed 0 blocks and 0 bytes of code.
BOLT-INFO: running hfsort+ for 472 functions
BOLT-INFO: program-wide dynostats after all optimizations before SCTC and FOP (no change):

                   0 : executed forward branches
                   0 : taken forward branches
                   0 : executed backward branches
                   0 : taken backward branches
                   0 : executed unconditional branches
                   0 : all function calls
                   0 : indirect calls
                   0 : PLT calls
                   0 : executed instructions
                   0 : executed load instructions
                   0 : executed store instructions
                   0 : taken jump table branches
                   0 : taken unknown indirect branches
                   0 : total branches
                   0 : taken branches
                   0 : non-taken conditional branches
                   0 : taken conditional branches
                   0 : all conditional branches

BOLT-INFO: SCTC: patched 152 tail calls (149 forward) tail calls (3 backward) from a total of 152 while removing 0 double jumps and removing 128 basic blocks totalling 640 bytes of code. CTCs total execution count is 0 and the number of times CTCs are taken is 0.
BOLT-INFO: padding code to 0x5800000 to accommodate hot text
BOLT-INFO: setting _end to 0x2e42464
BOLT-INFO: setting _end to 0x2e42464
BOLT-INFO: setting __hot_start to 0x4200000
BOLT-INFO: setting __hot_end to 0x5798dce
BOLT-INFO: Wrote 50178 BAT maps
BOLT-INFO: Wrote 0 BAT cold-to-hot entries

**Comment by maksfb** [comment]
Did you re-run `perf2bolt`? Could you share the new output from it?

**Comment by zpget** [comment]
> Did you re-run `perf2bolt`? Could you share the new output from it?

ERF2BOLT: spawning perf job to read branch events
PERF2BOLT: spawning perf job to read mem events
PERF2BOLT: spawning perf job to read process events
PERF2BOLT: spawning perf job to read task events
BOLT-INFO: Target architecture: x86_64
BOLT-INFO: first alloc address is 0x400000
BOLT-INFO: creating new program header table at address 0x6000000, offset 0x5c00000
BOLT-INFO: Parsed 50616 BAT entries
BOLT-INFO: Parsed 438 BAT cold-to-hot entries
BOLT-INFO: disabling -align-macro-fusion in non-relocation mode
BOLT-INFO: pre-processing profile using perf data aggregator
BOLT-INFO: profile collection done on a binary already processed by BOLT
BOLT-WARNING: build-id will not be checked because we could not read one from input binary
PERF2BOLT: waiting for perf mmap events collection to finish...
wait result:0 2
wait end result:0
PERF2BOLT: parsing perf-script mmap events output
PERF2BOLT: waiting for perf task events collection to finish...
PERF2BOLT: parsing perf-script task events output
PERF2BOLT: input binary is associated with 1 PID(s)
PERF2BOLT: waiting for perf events collection to finish...
PERF2BOLT: parse branch events...
PERF2BOLT-WARNING: using Intel Skylake bug workaround
PERF2BOLT: read 10260 samples and 138985 LBR entries
PERF2BOLT: 0 samples (0.0%) were ignored
PERF2BOLT: traces mismatching disassembled function contents: 0 (0.0%)
PERF2BOLT: out of range traces involving unknown regions: 60887 (50.0%)
PERF2BOLT: processing branch events...
PERF2BOLT: wrote 5068 objects and 0 memory objects to bin_perf.fdata

**Comment by maksfb** [comment]
The output looks correct. Can you run `llvm-bolt` with `-p bin_perf.fdata` instead of `-b=...`?

**Comment by maksfb** [comment]
> The output looks correct. Can you run `llvm-bolt` with `-p bin_perf.fdata` instead of `-b=...`?

Passing non-BOLTed binary as the input.

**Comment by zpget** [comment]
> > The output looks correct. Can you run `llvm-bolt` with `-p bin_perf.fdata` instead of `-b=...`?
> 
> Passing non-BOLTed binary as the input.

use -p  test   now  success!!

llvm-bolt: WARNING: reading perf data directly is unsupported, please use -aggregate-only or perf2bolt.
!!! Proceed on your own risk. !!!
BOLT-INFO: Target architecture: x86_64
BOLT-INFO: first alloc address is 0x400000
BOLT-INFO: creating new program header table at address 0x4000000, offset 0x3c00000
BOLT-INFO: enabling relocation mode
BOLT-INFO: pre-processing profile using branch profile reader
BOLT-INFO: profile collection done on a binary already processed by BOLT
BOLT-INFO: forcing -jump-tables=move as PIC jump table was detected in function __cpu_indicator_init/1
BOLT-WARNING: 5 collisions detected while hashing binary objects. Use -v=1 to see the list.
BOLT-INFO: 600 out of 68829 functions in the binary (0.9%) have non-empty execution profile
BOLT-INFO: 5 functions with profile could not be optimized
BOLT-INFO: profile for 82 objects was ignored
BOLT-INFO: the input contains 5809 (dynamic count : 51) opportunities for macro-fusion optimization that are going to be fixed
BOLT-INFO: removed 264 'repz' prefixes with estimated execution count of 0 times.
BOLT-INFO: ICF folded 19356 out of 69534 functions in 12 passes. 1 functions had jump tables.
BOLT-INFO: Removing all identical functions will save 1391.48 KB of code space. Folded functions were called 0 times based on profile.
BOLT-INFO: basic block reordering modified layout of 282 (0.56%) functions
BOLT-INFO: UCE removed 0 blocks and 0 bytes of code.
BOLT-INFO: splitting separates 60809 hot bytes from 187861 cold bytes (24.45% of split functions is hot).
BOLT-INFO: running hfsort+ for 605 functions
BOLT-INFO: program-wide dynostats after all optimizations before SCTC and FOP:

               39557 : executed forward branches
               20293 : taken forward branches
               15113 : executed backward branches
                9258 : taken backward branches
                6809 : executed unconditional branches
               38339 : all function calls
               15441 : indirect calls
                5926 : PLT calls
              561157 : executed instructions
              147841 : executed load instructions
               90794 : executed store instructions
                   3 : taken jump table branches
                   0 : taken unknown indirect branches
               61479 : total branches
               36360 : taken branches
               25119 : non-taken conditional branches
               29551 : taken conditional branches
               54670 : all conditional branches

               43647 : executed forward branches (+10.3%)
                 418 : taken forward branches (-97.9%)
               11023 : executed backward branches (-27.1%)
                7667 : taken backward branches (-17.2%)
                3607 : executed unconditional branches (-47.0%)
               38339 : all function calls (=)
               15441 : indirect calls (=)
                5926 : PLT calls (=)
              560231 : executed instructions (-0.2%)
              147841 : executed load instructions (=)
               90794 : executed store instructions (=)
                   3 : taken jump table branches (=)
                   0 : taken unknown indirect branches (=)
               58277 : total branches (-5.2%)
               11692 : taken branches (-67.8%)
               46585 : non-taken conditional branches (+85.5%)
                8085 : taken conditional branches (-72.6%)
               54670 : all conditional branches (=)

BOLT-INFO: SCTC: patched 152 tail calls (151 forward) tail calls (1 backward) from a total of 152 while removing 0 double jumps and removing 128 basic blocks totalling 640 bytes of code. CTCs total execution count is 2008 and the number of times CTCs are taken is 0.
BOLT-INFO: setting _end to 0x2e42464
BOLT-INFO: setting _end to 0x2e42464
BOLT-INFO: setting __hot_start to 0x4200000
BOLT-INFO: setting __hot_end to 0x4215fff
BOLT-INFO: Wrote 50520 BAT maps
BOLT-INFO: Wrote 342 BAT cold-to-hot entries
end 2020-08-05 12:11:46 cost 68 s

**Comment by zpget** [comment]
> > The output looks correct. Can you run `llvm-bolt` with `-p bin_perf.fdata` instead of `-b=...`?
> 
> Passing non-BOLTed binary as the input.

1. has warning

llvm-bolt: WARNING: reading perf data directly is unsupported, please use -aggregate-only or perf2bolt.
!!! Proceed on your own risk. !!!

2. use BOLTed   collect profile , BOLTed bin parse fdata,   non-BOLTed binary gen new BOLTed bin, Is there any risk? or loss  some  effect of optimization ?

**Comment by maksfb** [comment]
Sorry, I meant for you to try `-data bin_perf.fdata`. The end effect should be the same, but there would be no warning.

**Comment by zpget** [comment]
> Sorry, I meant for you to try `-data bin_perf.fdata`. The end effect should be the same, but there would be no warning.

now use -data bin_perf.fdata has no waring
 'llvm-bolt: WARNING: reading perf data directly is unsupported, please use -aggregate-only or perf2bolt.
!!! Proceed on your own risk. !!!'

method1:

in order to save the cost of collecting data, in real production environment collection profile data

STEP1.  first use bolt,  use  non-BOLTed collect profile , non-BOLTed bin parse fdata, non-BOLTed binary gen new BOLTed bin
STEP2.  use BOLTed(in real production environment) collect profile , BOLTed bin parse fdata, non-BOLTed binary gen new BOLTed bin,  and continue STEP2 ....


method2:
offline  environment
STEP1.  first use bolt,  use  non-BOLTed collect profile , non-BOLTed bin parse fdata, non-BOLTed binary gen new BOLTed bin, and continue STEP2 ....


method1 VS method2

Confirm method1  is  there any risk ? or loss some effect of optimization vs method2 ?

method1 can I use
STEP2:
use BOLTed(in real production environment) collect profile , non-BOLTed bin parse fdata, non-BOLTed binary gen new BOLTed bin,  and continue STEP2 ....

**Comment by maksfb** [comment]
If I understand correctly, the 1st approach will work. It has the advantage that you do not need any special deployment. Since you are talking about continuous deployment and optimizations, there will be changes in the source code and often the binary that you've collected profile on is not the same as the one you are trying to optimize because of the source code has changed. In that case, you will get some percentage of stale functions and will lose effectiveness of code layout optimizations. Hard to say how much without knowing specifics.

To achieve 100% efficiency of BOLT optimizations, you have to collect the profile on the same binary that you are going to optimize. It could be BOLTed, but the source code revision has to be identical.

TL;DR: if you are using BAT mode, it doesn't matter if you collect profile on BOLTed or non-BOLTed binary. It's more important for the profiled binary (or pre-BOLTed profiled binary) to match the one that you optimize with `llvm-bolt`.

**Comment by zpget** [comment]
> If I understand correctly, the 1st approach will work. It has the advantage that you do not need any special deployment. Since you are talking about continuous deployment and optimizations, there will be changes in the source code and often the binary that you've collected profile on is not the same as the one you are trying to optimize because of the source code has changed. In that case, you will get some percentage of stale functions and will lose effectiveness of code layout optimizations. Hard to say how much without knowing specifics.
> 
> To achieve 100% efficiency of BOLT optimizations, you have to collect the profile on the same binary that you are going to optimize. It could be BOLTed, but the source code revision has to be identical.
> 
> TL;DR: if you are using BAT mode, it doesn't matter if you collect profile on BOLTed or non-BOLTed binary. It's more important for the profiled binary (or pre-BOLTed profiled binary) to match the one that you optimize with `llvm-bolt`.

thank you very much~
In BAT mode, collect profile on BOLTed, trans perf.data to perf.fdata must use BOLTed binary,
I try use non-BOLTed binary trans perf.data to perf.fdata ,  use llvm-bolt gen new BOLTed bin has SCTC and FOP (no change)
BOLT-INFO: program-wide  dynostats after all optimizations before SCTC and FOP (no change):

                   0 : executed forward branches
                   0 : taken forward branches
                   0 : executed backward branches
                   0 : taken backward branches
                   0 : executed unconditional branches
                   0 : all function calls
                   0 : indirect calls
                   0 : PLT calls
                   0 : executed instructions
                   0 : executed load instructions
                   0 : executed store instructions
                   0 : taken jump table branches
                   0 : taken unknown indirect branches
                   0 : total branches
                   0 : taken branches
                   0 : non-taken conditional branches
                   0 : taken conditional branches
                   0 : all conditional branches

**Comment by maksfb** [comment]
Right. `perf2bolt` has to run on the binary that was used to generate `perf.data`.
